coursier.bzl

```diff
@@ -432,23 +461,132 @@ def _get_java_proxy_args(repository_ctx):
-            "-lc",
-            "cat " + str(repository_ctx.path(filepath)),
-        ])
-    else:
-        exec_result = repository_ctx.execute([
-            repository_ctx.which("cat"),
-            repository_ctx.path(filepath),
-        ])
-    if (exec_result.return_code != 0):
-        fail("Error while trying to read %s: %s" % (filepath, exec_result.stderr))
-    return exec_result.stdout
+        bash = repository_ctx.os.environ.get("BAZEL_SH")
+        if (bash == None):
+            fail("Please set the BAZEL_SH environment variable to the path of MSYS2 bash. " +
+                 "This is typically `c:\\msys64\\usr\\bin\\bash.exe`. For more information, read " +
+                 "https://docs.bazel.build/versions/master/install-windows.html#getting-bazel")
+        repository_ctx.execute([bash, "-lc", "echo", "works"])
+
+def _pinned_coursier_fetch_impl(repository_ctx):
+    if not repository_ctx.attr.maven_install_json:
+        fail("Please specify the file label to maven_install.json (e.g."
+             + "//:maven_install.json).")
+
+    _windows_check(repository_ctx)
+
+    artifacts = []
+    for a in repository_ctx.attr.artifacts:
+        artifacts.append(json_parse(a))
+
+    # Read Coursier state from maven_install.json.
+    repository_ctx.symlink(
+        repository_ctx.path(repository_ctx.attr.maven_install_json),
+        repository_ctx.path("imported_maven_install.json")
+    )
+    maven_install_json_content = json_parse(
+        repository_ctx.read(
+            repository_ctx.path("imported_maven_install.json")),
+        fail_on_invalid = False,
+    )
+    if maven_install_json_content == None:
+        fail("Failed to parse %s. Is this file valid JSON? The file may have been corrupted." % repository_ctx.path(repository_ctx.attr.maven_install_json)
+             + "Consider regenerating maven_install.json with the following steps:\n"
+             + "  1. Remove the maven_install_json attribute from your `maven_install` declaration for `@%s`.\n" % repository_ctx.name
+             + "  2. Regenerate `maven_install.json` by running the command: bazel run @%s//:pin" % repository_ctx.name
+             + "  3. Add `maven_install_json = \"//:maven_install.json\"` into your `maven_install` declaration.")
+
+    if maven_install_json_content.get("dependency_tree") == None:
+        fail("Failed to parse %s. " % repository_ctx.path(repository_ctx.attr.maven_install_json)
+                + "It is not a valid maven_install.json file. Has this "
+                + "file been modified manually?")
+
+    dep_tree = maven_install_json_content["dependency_tree"]
+
+    # Create the list of http_file repositories for each of the artifacts
+    # in maven_install.json. This will be loaded additionally like so:
+    #
+    # load("@maven//:defs.bzl", "pinned_maven_install")
+    # pinned_maven_install()
+    http_files = [
+        "load(\"@bazel_tools//tools/build_defs/repo:http.bzl\", \"http_file\")",
```

Comment:
Can we have a comment at the very top of this file to remind users that this is a generated file which shouldn't be modified manually?
The message could be like `# Do not edit. rules_jvm_external autogenerates this file from maven_install Please run ... command to update this file.`

Reply:
It's JSON, so no comments :-) 

I can turn it into a "private" field in the JSON object, through.

Reply:
Ah, I totally forget json has not comment. Thinking of a better way to remind users.

Reply:
How about changing `maven_install.json` to something more obvious that implies we shouldn't modify it manually? Like `maven_autogenerated.json`.

Reply:
I think that still doesn't imply that you shouldn't modify it manually. Do you think users will end up modifying this manually often enough that it becomes a user experience problem?

Reply:
I am thinking of a scenario that someone may think changing the artifacts version is as simple as modifying `maven_install.json`, but it may break other artifacts which have the dependency on the changed artifact.

Since I am going to introduce `rules_jvm_external` to the entire organization, there is a chance someone doesn't notice the proper way of using it.

Reply:
Perhaps we can hash the contents of the `dependencies` array, and add that as a field to the generated object? Then `pinned_coursier_maven` would compare the hash before generating the `http_file` targets and see if it was modified manually, and ask the user to not do that.

Reply:
It sounds like a good idea to me.

Reply:
I'll add this in a follow up PR.